src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

```diff
@@ -264,19 +264,32 @@ SelectPartsDecision MergeTreeDataMergerMutator::selectPartsToMerge(
             if (!can_merge_callback(nullptr, part, nullptr))
                 continue;
 
+            /// This part can be merged only with next parts (no prev part exists), so start
+            /// new interval if previous was not empty.
+            if (!parts_ranges.back().empty())
+                parts_ranges.emplace_back();
         }
         else
         {
             /// If we cannot merge with previous part we had to start new parts
             /// interval (in the same partition)
             if (!can_merge_callback(*prev_part, part, nullptr))
             {
-                /// Starting new interval in the same partition
+                /// Now we have no previous part
+                prev_part = nullptr;
+
+                /// Mustn't be empty
                 assert(!parts_ranges.back().empty());
-                parts_ranges.emplace_back();
 
-                /// Now we have no previous part, but it affects only logging
-                prev_part = nullptr;
+                /// Some parts cannot be merged with previous parts and also cannot be merged with themselves,
+                /// for example, merge is already assigned for such parts, or they participate in quorum inserts
+                /// and so on.
+                /// Also we don't start new interval here (maybe all next parts cannot be merged and we don't want to have empty interval)
+                if (!can_merge_callback(nullptr, part, nullptr))
```

Comment:
Do we need to start new interval in this case?

Reply:
Ok, seems like it's done here
https://github.com/ClickHouse/ClickHouse/blob/c15493f0df3dc577e22760553eff452c07a83fba/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp#L269-L270